### PR TITLE
Exiles Reach Down with the Quilboar and Forbidden Quilboar Necromancy Quest

### PR DESCRIPTION
**Pull request by ghost**
Added Quest Down with the Quilboar and Forbidden Quilboar Necromancy for Alliance and Horde.

Bring on the plains!!!!

**Issues addressed:**

**Tests performed:**

(Does it build, tested in-game, etc.)
Builds and tested.

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] Not sure about the phase I used for areatrigger but it has to be common to both sides. Leaving that to TC Devs to deteremine.
- [ ] There is a hack to allow Geolord conversation. Leaving that to TC Devs.

### CODE REVIEW

**Comment by malcrom** [comment]
No idea how to fix my arch nemisis nopch.

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 1]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+struct npc_geolord_grekog : public ScriptedAI
+{
+    npc_geolord_grekog(Creature* creature) : ScriptedAI(creature) { }
+
+    void JustAppeared() override
+    {
+        uint32 _entry = NPC_LINDIE_SPRINGSTOCK;
+        std::string_view creatureString = "briarpatch_bunny_alliance";
+
+        if (me->GetInstanceId())
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4454]: i dont think this is the correct check we want here; check for worldstate to know if its alliance/horde map; just like we are doing it for the spawngroups in conditions

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 2]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+            _entry = NPC_CORK_FIZZLEPOP;
+            creatureString = "Briarpatch_bunny_horde";
+        }
+
+        Creature* bunny = FindCreatureIgnorePhase(me, creatureString, 25.0f);
+
+        if (!bunny)
+            return;
+
+        Creature* prisoner = bunny->SummonCreature(_entry, 16.4271f, -2511.82f, 78.8215f, 5.66398f, TEMPSUMMON_MANUAL_DESPAWN);
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4465]: move coords to const variable

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 3]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+        _prisonerGUID = prisoner->GetGUID();
+    }
+
+    void JustDied(Unit* /*killer*/) override
+    {
+        if (Creature* prisoner = ObjectAccessor::GetCreature(*me, _prisonerGUID))
+            prisoner->AI()->SetData(SET_DATA_FREE_PRISONER, SET_DATA_FREE_PRISONER);
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4472]: we have AI()->DoAction in C++ for such purposes, SetData only if you have to pass values which may vary

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 4]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+
+    SET_DATA_FREE_PRISONER     = 1,
+
+    SPELL_NECROTIC_RITUAL_DNT  = 305513,
+    SPELL_EARTH_BOLT           = 270453,
+    SPELL_UPHEAVAL             = 319273
+};
+
+// This script is used by Geolord Grek'og for Down with the Quilboar quest
+struct npc_geolord_grekog : public ScriptedAI
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4445]: put comment above `struct npc_geolord_[...]` with following syntax: `// <id> - <name>`, do this for all creature scripts, if multiple ids are involved use a new line for each - like:
```c
// <id a> - <npc name a>
// <id b> - <npc name b>
struct npc_awesome [...]
```
You may keep ur comment `// This script is used by Geolord Grek'og for Down with the Quilboar quest` additionally if you want :)

i'm only commenting it here, you should do that for all npc scripts below aswell

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 5]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+        me->SetDisableGravity(true);
+        me->SetTemplateRooted(true);
+        me->CastSpell(me, SPELL_NECROTIC_RITUAL_DNT);
+    }
+
+    void SetData(uint32 /*type*/, uint32 data) override
+    {
+        if (data == SET_DATA_FREE_PRISONER)
+        {
+            Position jumpto = { 19.5174f, -2513.75f, 74.0545f, me->GetOrientation() };
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4552]: move to const variable, i kinda assume the orientation does not change due to being rooted

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 6]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+        _outOfCombat = false;
+
+        if (me->HasAura(SPELL_QUILBOAR_SLEEP_DNT))
+        {
+            me->SetReactState(REACT_AGGRESSIVE);
+            me->RemoveAura(SPELL_QUILBOAR_SLEEP_DNT);
+            me->SetStandState(UNIT_STAND_STATE_STAND);
+        }
+
+        if (!urand(0, 2))
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4637]: `if (roll_chance_f(33.33f))`

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 7]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+            Talk(SAY_AGGRO, who);
+
+        _events.ScheduleEvent(EVENT_BRUTAL_STRIKE, 3s, 5s);
+    }
+
+    void JustDied(Unit* killer) override
+    {
+        if (!urand(0, 2))
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4645]: `if (roll_chance_f(33.33f))`

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 8]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+            Talk(SAY_DEATH, killer);
+
+        for (auto const& [playerGuid, loot] : me->m_personalLoot)
+        {
+            for (LootItem const& lootItem : loot->items)
+            {
+                if (lootItem.itemid == ITEM_STITCHED_CLOTH_TUNIC || lootItem.itemid == ITEM_STITCHED_LEATHER_TUNIC || lootItem.itemid == ITEM_LINKED_MAIL_HAUBERK || lootItem.itemid == ITEM_DENTED_CHESTPLATE)
+                    if (Player* player = ObjectAccessor::GetPlayer(*me, playerGuid))
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4653]: move this check before `for (LootItem const& lootItem : loot->items)`, as we do not need to retrieve it multiple times, once is enough

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 9]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+            return;
+
+        _events.Update(diff);
+
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
+            {
+                case EVENT_CHECK_FOR_AURA:
+                    if (me->HasAura(SPELL_QUILBOAR_SLEEP_DNT))
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4671]: Given the fact that the aura in question is having `ChannelInterrupt Flags: 0x1000100E (Damage, Action, Moving, Attacking, EnteringCombat),` we can just use `AfterEffectApply` to set reactstate and `AfterEffectRemove` to remove it, set standstate and other stuff

=> implement it in aurascript for `313265 Quilboar Sleep (DNT)`, so we do not need it for every quilboar separately

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 10]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+        _outOfCombat = false;
+
+        if (me->HasAura(SPELL_QUILBOAR_SLEEP_DNT))
+        {
+            me->SetReactState(REACT_AGGRESSIVE);
+            me->RemoveAura(SPELL_QUILBOAR_SLEEP_DNT);
+            me->SetStandState(UNIT_STAND_STATE_STAND);
+        }
+
+        if (!urand(0, 2))
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4722]: `if (roll_chance_f(33.33f))`

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 11]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+            Talk(SAY_AGGRO, who);
+
+        _events.ScheduleEvent(EVENT_GEOMANCER_EARTH_BOLT, 3s, 5s);
+    }
+
+    void JustDied(Unit* killer) override
+    {
+        if (!urand(0, 2))
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4730]: `if (roll_chance_f(33.33f))`

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 12]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+    void JustDied(Unit* killer) override
+    {
+        Talk(SAY_DEATH, killer);
+
+        for (auto const& [playerGuid, loot] : me->m_personalLoot)
+        {
+            for (LootItem const& lootItem : loot->items)
+            {
+                if (lootItem.itemid == ITEM_BATTERED_CLOAK || lootItem.itemid == ITEM_OVERSEERS_MANDATE)
+                    if (Player* player = ObjectAccessor::GetPlayer(*me, playerGuid))
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4815]: same here, move player above `for (LootItem [...])`

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 13]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+    {
+        if (!urand(0, 2))
+            Talk(SAY_DEATH, killer);
+
+        for (auto const& [playerGuid, loot] : me->m_personalLoot)
+        {
+            for (LootItem const& lootItem : loot->items)
+            {
+                if (lootItem.itemid == ITEM_STITCHED_CLOTH_TUNIC || lootItem.itemid == ITEM_STITCHED_LEATHER_TUNIC || lootItem.itemid == ITEM_LINKED_MAIL_HAUBERK || lootItem.itemid == ITEM_DENTED_CHESTPLATE)
+                    if (Player* player = ObjectAccessor::GetPlayer(*me, playerGuid))
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4738]: same here, move player above `for (LootItem [...])`

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp [hunk 14]
```diff
@@ -4114,6 +4296,616 @@ class spell_tutorial_health_dnt : public SpellScript
...
+                        player->SetRewardedQuest(QUEST_BRIARPATCH_CHEST_DROPPED);
```
Comment (mdX7) [inline src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4654]: we can also `break;` here, as further iterations are no longer needed for the loop (you may also check if quest is rewarded below this loop and break outer loop aswell)

File: sql/updates/world/master/9999_99_99_99_world.sql [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+SET @ID := 65; -- One required
+SET @SPAWNID := 65; -- One required
+SET @CREATEPROPID := 65; -- One required
+
+-- Briar Patch
+
+-- Add missing spawns
+DELETE FROM `creature` WHERE `guid` IN (@GUID,@GUID+1);
+INSERT INTO `creature` (`guid`, `id`, `map`, `zoneId`, `areaId`, `spawnDifficulties`, `phaseUseFlags`, `PhaseId`, `PhaseGroup`, `terrainSwapMap`, `modelid`, `equipment_id`, `position_x`, `position_y`, `position_z`, `orientation`, `spawntimesecs`, `wander_distance`, `currentwaypoint`, `curhealth`, `curmana`, `MovementType`, `npcflag`, `unit_flags`, `unit_flags2`, `unit_flags3`, `ScriptName`, `StringId`, `VerifiedBuild`) VALUES
+(@GUID,155371,2175,10424,10503,0,0,15314,0,-1,0,0,9.38368,-2514.89,75.9253,0,7200,0,0,124,0,0,NULL,NULL,NULL,NULL,'','Briarpatch_bunny_horde',52649),
```
[the remark this answers is not in the document]
Comment (malcrom) [reply sql/updates/world/master/9999_99_99_99_world.sql:14]: fixed.

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 2 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4465]: Fixed

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 4 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4445]: Fixed

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 7 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4645]: fixed

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 10 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4722]: fixed

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 6 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4637]: fixed

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 11 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4730]: fixed

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 14 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4654]: Fixed

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 1 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4460]: Don't know how to do it.

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 3 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4478]: Don't know how to do it.

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 9 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4687]: Will look into aurascript later today.

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 1 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp]: fixed

File: src/server/scripts/ExilesReach/zone_exiles_reach.cpp (hunk 3 above)
Comment (malcrom) [reply src/server/scripts/ExilesReach/zone_exiles_reach.cpp:4481]: fixed

**Comment by mdX7** [comment]
depends on #29552

**Pull request merged**